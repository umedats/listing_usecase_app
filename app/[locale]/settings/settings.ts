/**
 * Object which will store the necessary settings required to make SP API calls.
 * Each field in this object corresponds to a field in the settings page.
 *
 * region: AWS Region for creating infrastructure
 * accountId: AWS AccountId for creating infrastructure
 * clientId: Client Id part of the selling partner application LWA credentials
 * clientSecret: Client Secret part of the selling partner application LWA credentials
 * refreshToken: Refresh Token used to get the access token
 * sellingPartnerId: Selling Partner Id
 * sellingPartnerIdType: A selling partner identifier type
 * marketplaceId: The MarketplaceId to be used during the Listing Process
 */
export type Settings = {
  region: string;
  accountId: string;
  clientId: string;
  clientSecret: string;
  refreshToken: string;
  sellingPartnerId: string;
  sellingPartnerIdType: string;
  marketplaceId: string;
};

/**
 * returns true if the settings are incomplete.
 * @param settings
 */
export function areSettingsIncomplete(settings: Settings) {
  return (
    !settings.region ||
    !settings.accountId ||
    !settings.clientId ||
    !settings.clientSecret ||
    !settings.refreshToken ||
    !settings.sellingPartnerId ||
    !settings.sellingPartnerIdType ||
    !settings.marketplaceId
  );
}

export const SELLING_PARTNER_TYPE_KEY_VENDOR_CODE = "Vendor Code";
export const SELLING_PARTNER_TYPE_KEY_MERCHANT_ACCOUNT_ID =
  "Merchant Account Id";
