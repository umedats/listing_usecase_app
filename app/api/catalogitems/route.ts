import { NextResponse } from "next/server";
import { headers } from "next/dist/client/components/headers";
import getSPAPIEndpoint, {
  buildCatalogItemsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  IDENTIFIER_TYPES_HEADER,
  IDENTIFIERS_HEADER,
  KEYWORDS_HEADER,
  SEARCH_CATALOG_ITEMS_API_DOC_LINK,
  SEARCH_CATALOG_ITEMS_API_NAME,
  LOCALE_HEADER,
} from "@/app/constants/global";
import nextResponse from "@/app/utils/next-response-factory";
import { Item, ItemSearchResults } from "@/app/sdk/catalogItems_2022-04-01";
import { ProductSearchResult } from "@/app/[locale]/create-offer/product-result-type";
import { Settings } from "@/app/[locale]/settings/settings";
import { SPAPIRequestResponse } from "@/app/model/types";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves the list of Catalog Items based on identifiers/keywords by calling the SearchCatalogItems SP-API.
 * @constructor
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getCatalogItemsHandler);
}

async function getCatalogItemsHandler(settings: Settings) {
  const { identifiersType, identifiers, keywords, locale } =
    retrieveRequestParamsFromHeader();
  const optionalParams = constructOptionalParametersForApi(
    settings,
    identifiersType,
    identifiers,
    keywords,
    locale,
  );
  const nextResponseOrSPAPIRequestResponse = await fetchCatalogItems(
    settings,
    optionalParams,
  );

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  return fetchItemSearchResults(nextResponseOrSPAPIRequestResponse);
}

function retrieveRequestParamsFromHeader() {
  const requestHeaders = headers();
  const identifiersType = requestHeaders.get(IDENTIFIER_TYPES_HEADER);
  const identifierData = requestHeaders.get(IDENTIFIERS_HEADER);
  const identifiers = identifierData ? identifierData.split(",") : null;
  const keywordsData = requestHeaders.get(KEYWORDS_HEADER);
  const keywords = keywordsData ? keywordsData.split(",") : null;
  const locale = requestHeaders.get(LOCALE_HEADER);

  return { identifiersType, identifiers, keywords, locale };
}

function constructOptionalParametersForApi(
  settings: Settings,
  identifiersType: string | null,
  identifiers: string[] | null,
  keywords: string[] | null,
  locale: string | null,
) {
  const includedData = ["productTypes", "summaries"];

  if (identifiers) {
    const optionalParametersForIdentifiers = {
      identifiers: identifiers,
      identifiersType: identifiersType,
      includedData: includedData,
      locale: locale,
    };

    // Seller ID is mandatory when identifier Type is SKU
    if (identifiersType === "SKU") {
      return {
        sellerId: settings.sellingPartnerId,
        ...optionalParametersForIdentifiers,
      };
    }

    // For non-SKU Identifier Type use case.
    return optionalParametersForIdentifiers;
  }

  if (keywords) {
    return {
      keywords: keywords,
      includedData: includedData,
      locale: locale,
    };
  }
}

async function fetchCatalogItems(settings: Settings, opts: any) {
  // Build Catalog Items API Client.
  const catalogApi = await buildCatalogItemsAPIClient(
    getSPAPIEndpoint(settings.region),
  );

  const reqResponse = {
    apiName: SEARCH_CATALOG_ITEMS_API_NAME,
    apiDocumentationLink: SEARCH_CATALOG_ITEMS_API_DOC_LINK,
    request: {
      marketplaceIds: [settings.marketplaceId],
      ...opts,
    },
    response: {},
  };
  try {
    reqResponse.response = await catalogApi.searchCatalogItems(
      reqResponse.request.marketplaceIds,
      opts,
    );
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}

function deriveTitle(item: Item) {
  return item.summaries?.[0]?.itemName;
}

function derivePTD(item: Item) {
  return item.productTypes?.[0]?.productType;
}

function fetchItemSearchResults(reqResponse: SPAPIRequestResponse) {
  const searchResults: ItemSearchResults =
    ItemSearchResults.constructFromObject(reqResponse.response, undefined);

  if (searchResults.numberOfResults && searchResults.numberOfResults > 0) {
    // @ts-ignore
    const items: Item[] = searchResults.items;
    const results: ProductSearchResult[] = items.map((item, index) => {
      const searchResult: ProductSearchResult = {
        title: deriveTitle(item),
        asin: item.asin,
        productType: derivePTD(item),
      };
      return searchResult;
    });
    return nextResponse(
      200,
      "OK",
      serializeToJsonString({
        data: results,
        debugContext: [reqResponse],
      }),
    );
  }

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: [],
      debugContext: [reqResponse],
    }),
  );
}
