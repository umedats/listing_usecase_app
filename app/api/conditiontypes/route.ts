import nextResponse from "@/app/utils/next-response-factory";
import { Settings } from "@/app/[locale]/settings/settings";
import { NextResponse } from "next/server";
import { headers } from "next/dist/client/components/headers";
import {
  ASIN_HEADER,
  GET_LISTING_RESTRICTIONS_API_DOC_LINK,
  GET_LISTING_RESTRICTIONS_API_NAME,
  LOCALE_HEADER,
} from "@/app/constants/global";
import getSPAPIEndpoint, {
  buildListingsRestrictionsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  Restriction,
  RestrictionList,
} from "@/app/sdk/listingsRestrictions_2021-08-01";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves the conditions under which a MCID can list a given ASIN by calling SPAPI.
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getConditionTypesHandler);
}

async function getConditionTypesHandler(settings: Settings) {
  // Validate the various headers expected by the API.
  const validateHeadersResponse = validateHeaderParams();
  if (validateHeadersResponse instanceof NextResponse) {
    return validateHeadersResponse;
  }

  const { asin, locale } = validateHeadersResponse;

  // Retrieve restricted Conditions from SP API
  const nextResponseOrSPAPIResponse = await fetchRestrictions(
    settings,
    asin,
    locale,
  );
  if (nextResponseOrSPAPIResponse instanceof NextResponse) {
    return nextResponseOrSPAPIResponse;
  }
  const restrictedConditionTypes: string[] =
    RestrictionList.constructFromObject(
      nextResponseOrSPAPIResponse.response,
      undefined,
    ).restrictions.map((r: Restriction) => r.conditionType);

  const allowedTypes = Object.values(Restriction.ConditionTypeEnum).filter(
    (c) => !restrictedConditionTypes.includes(c),
  );
  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: allowedTypes,
      debugContext: [nextResponseOrSPAPIResponse],
    }),
  );
}

// Validates the header parameters.
function validateHeaderParams() {
  const requestHeaders = headers();
  const asin = requestHeaders.get(ASIN_HEADER);
  if (!asin) {
    const errorMessage = "The request is missing the asin in the headers";
    console.log(errorMessage);
    return nextResponse(400, errorMessage);
  }

  const locale = requestHeaders.get(LOCALE_HEADER);

  return { asin, locale };
}

/**
 * fetches the condition types from the SP API.
 * @param settings user entered settings
 * @param asin asin
 * @param locale user preferred locale.
 */
async function fetchRestrictions(
  settings: Settings,
  asin: string,
  locale: string | null,
) {
  // Build the Restrictions API Client.
  const restrictionsApi = await buildListingsRestrictionsAPIClient(
    getSPAPIEndpoint(settings.region),
  );

  const reqResponse = {
    apiName: GET_LISTING_RESTRICTIONS_API_NAME,
    apiDocumentationLink: GET_LISTING_RESTRICTIONS_API_DOC_LINK,
    request: {
      asin: asin,
      sellerId: settings.sellingPartnerId,
      marketplaceIds: [settings.marketplaceId],
      conditionType: null,
      reasonLocale: locale ? locale : "",
    },
    response: {},
  };

  try {
    reqResponse.response = await restrictionsApi.getListingsRestrictions(
      reqResponse.request.asin,
      reqResponse.request.sellerId,
      reqResponse.request.marketplaceIds,
      {
        conditionType: reqResponse.request.conditionType,
        reasonLocale: reqResponse.request.reasonLocale,
      },
    );
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}
