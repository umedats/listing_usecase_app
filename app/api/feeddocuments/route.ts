import nextResponse from "@/app/utils/next-response-factory";
import { serializeToJsonString } from "@/app/utils/serialization";
import { NextResponse } from "next/server";
import { Settings } from "@/app/[locale]/settings/settings";
import { headers } from "next/dist/client/components/headers";
import {
  CONTENT_TYPE_HEADER,
  FEED_DOCUMENT_ID_HEADER,
  GET_FEED_DOCUMENT_API_NAME,
  GET_FEED_DOCUMENT_DOC_LINK,
  NO_CACHING_DIRECTIVE,
} from "@/app/constants/global";
import getSPAPIEndpoint, {
  buildFeedsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import { FeedDocument } from "@/app/sdk/feeds_2021-06-30";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";
import { unzipSync } from "zlib";
import { SPAPIRequestResponse } from "@/app/model/types";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves a feed document content and returns the content in the API response.
 * @constructor
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getFeedDocumentContentHandler);
}

async function getFeedDocumentContentHandler(settings: Settings) {
  const requiredParametersResponse =
    await validateAndRetrieveRequiredParametersForGetRequest();

  if (requiredParametersResponse instanceof NextResponse) {
    return requiredParametersResponse;
  }

  const { feedDocumentId } = requiredParametersResponse;

  const nextResponseOrSPAPIRequestResponse = await getFeedDocumentUrl(
    settings,
    feedDocumentId,
  );

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  const getFeedDocumentContentResponse = await getFeedDocumentContent(
    FeedDocument.constructFromObject(
      nextResponseOrSPAPIRequestResponse.response,
      undefined,
    ),
    nextResponseOrSPAPIRequestResponse,
  );

  if (getFeedDocumentContentResponse instanceof NextResponse) {
    return getFeedDocumentContentResponse;
  }

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: {
        content: getFeedDocumentContentResponse.feedDocumentContent,
        contentType: getFeedDocumentContentResponse.feedDocumentContentType,
      },
      debugContext: [nextResponseOrSPAPIRequestResponse],
    }),
  );
}

async function validateAndRetrieveRequiredParametersForGetRequest() {
  const validateHeadersResponse =
    validateAndRetrieveRequestParamsFromHeaderForGetRequest();
  if (validateHeadersResponse instanceof NextResponse) {
    return validateHeadersResponse;
  }

  const { feedDocumentId } = validateHeadersResponse;
  return { feedDocumentId };
}

function validateAndRetrieveRequestParamsFromHeaderForGetRequest() {
  const requestHeaders = headers();
  const feedDocumentId = requestHeaders.get(FEED_DOCUMENT_ID_HEADER);
  if (!feedDocumentId) {
    const errorMessage =
      "The request is missing the feedDocumentId in the headers";
    return nextResponse(400, errorMessage);
  }

  return { feedDocumentId };
}

async function getFeedDocumentUrl(settings: Settings, feedDocumentId: string) {
  const feedsApi = await buildFeedsAPIClient(getSPAPIEndpoint(settings.region));
  const reqResponse = {
    apiName: GET_FEED_DOCUMENT_API_NAME,
    apiDocumentationLink: GET_FEED_DOCUMENT_DOC_LINK,
    request: {
      feedDocumentId: [feedDocumentId],
    },
    response: {},
  };

  try {
    reqResponse.response = await feedsApi.getFeedDocument(feedDocumentId);
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}

async function getFeedDocumentContent(
  feedDocument: FeedDocument,
  requestResponse: SPAPIRequestResponse,
) {
  try {
    const fetchDocumentResponse = await fetch(feedDocument.url, {
      cache: NO_CACHING_DIRECTIVE,
    });
    const feedDocumentContentType =
      fetchDocumentResponse.headers.get(CONTENT_TYPE_HEADER);
    let feedDocumentContent = "";
    if (feedDocument.compressionAlgorithm) {
      feedDocumentContent = unzipSync(
        await fetchDocumentResponse.arrayBuffer(),
      ).toString();
    } else {
      feedDocumentContent = await fetchDocumentResponse.text();
    }
    return { feedDocumentContent, feedDocumentContentType };
  } catch (error) {
    // Generally these errors are recoverable.
    const errorMessage = "Error while fetching the feed document content.";
    console.log(errorMessage + " " + error);
    return nextResponse(
      500,
      errorMessage,
      serializeToJsonString({
        debugContext: [requestResponse],
      }),
    );
  }
}
