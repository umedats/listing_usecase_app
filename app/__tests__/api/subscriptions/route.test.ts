import {
  createSubscription,
  deleteSubscription,
  getDisplayableListingSubscriptions,
} from "@/app/utils/subscription";
import { MOCK_SETTINGS } from "@/app/test-utils/mock-settings";
import { getSettings } from "@/app/api/settings/wrapper";
import fetch from "jest-fetch-mock";
import {
  MOCK_CREATE_SUBSCRIPTION_SP_API_RESPONSE,
  MOCK_SUBSCRIPTION_1,
  MOCK_SUBSCRIPTION_2,
  MOCK_SUBSCRIPTION_3,
  MOCK_SUBSCRIPTION_4,
} from "@/app/test-utils/mock-subscription";
import { verifyStatusAndBody } from "@/app/test-utils/next-response-helper";
import { GET, DELETE, POST } from "@/app/api/subscriptions/route";
import { NextRequest } from "next/server";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  removeTargetAndDeleteRule,
  createEventBridgeResources,
} from "@/app/utils/eventbridge";
import { deleteQueue } from "@/app/utils/sqs";
import {
  createEventBridgeDestination,
  getEventBridgeDestinations,
} from "@/app/utils/destination";
import {
  MOCK_CREATE_DESTINATION_RESPONSE,
  MOCK_DESTINATION,
} from "@/app/test-utils/mock-destination";
import { headers } from "next/dist/client/components/headers";
import { afterEach } from "@jest/globals";
import { instance, reset, when, mock } from "ts-mockito";
import { NOTIFICATION_TYPE_HEADER } from "@/app/constants/global";
import { ReadonlyHeaders } from "next/dist/server/web/spec-extension/adapters/headers";

// @ts-ignore
const mockReadOnlyHeaders: ReadonlyHeaders = mock(ReadonlyHeaders);

jest.mock("next/dist/client/components/headers");
const mockedHeaders = jest.mocked(headers);

jest.mock("@/app/utils/eventbridge");
const mockedRemoveTargetAndDeleteRule = jest.mocked(removeTargetAndDeleteRule);
const mockedCreateEventBridgeResources = jest.mocked(
  createEventBridgeResources,
);
jest.mock("@/app/utils/sqs");
const mockedDeleteQueue = jest.mocked(deleteQueue);

jest.mock("@/app/utils/subscription");
const mockedGetDisplayableListingSubscriptions = jest.mocked(
  getDisplayableListingSubscriptions,
);
const mockedDeleteSubscription = jest.mocked(deleteSubscription);
jest.mock("@/app/utils/destination");
const mockedGetEventBridgeDestination = jest.mocked(getEventBridgeDestinations);
const mockedCreateEventBridgeDestination = jest.mocked(
  createEventBridgeDestination,
);
const mockedCreateSubscription = jest.mocked(createSubscription);
const LISC_NOTIFICATION = "LISTINGS_ITEM_STATUS_CHANGE";

jest.mock("@/app/api/settings/wrapper");
const mockedGetSettings = jest.mocked(getSettings);
function mockSettings(status: number) {
  mockedGetSettings.mockResolvedValue({
    status: status,
    statusText: "Some Status Text",
    settings: MOCK_SETTINGS,
  });
}

function mockHeaders(notificationType?: string) {
  notificationType &&
    when(mockReadOnlyHeaders.get(NOTIFICATION_TYPE_HEADER)).thenReturn(
      LISC_NOTIFICATION,
    );
  mockedHeaders.mockReturnValue(instance(mockReadOnlyHeaders));
}

describe("Test for the subscriptions API", () => {
  describe("Test for the GET subscriptions API", () => {
    beforeEach(() => {
      fetch.resetMocks();
    });

    test("returns 200 on successful case", async () => {
      mockSettings(200);
      mockedGetDisplayableListingSubscriptions.mockResolvedValue([
        MOCK_SUBSCRIPTION_1,
        MOCK_SUBSCRIPTION_2,
      ]);

      await verifyStatusAndBody(await GET(), 200, true);
    });

    test("returns 500 on settings failure", async () => {
      mockSettings(400);
      await verifyStatusAndBody(await GET(), 500, false);
    });

    test("returns 409 on runtime error", async () => {
      mockSettings(200);
      mockedGetDisplayableListingSubscriptions.mockRejectedValue({
        status: 409,
      });
      await verifyStatusAndBody(await GET(), 409, true);
    });
  });

  describe("Test for the DELETE subscriptions API", () => {
    beforeEach(() => {
      fetch.resetMocks();
    });

    function buildDeleteNextRequest(body: any) {
      return new NextRequest("http://localhost:3000/api/subscriptions", {
        method: "DELETE",
        body: serializeToJsonString(body),
      });
    }

    test("returns 200 on successful case", async () => {
      mockSettings(200);
      mockedDeleteSubscription.mockImplementationOnce(() => Promise.resolve());
      mockedRemoveTargetAndDeleteRule.mockImplementationOnce(() =>
        Promise.resolve(),
      );
      mockedDeleteQueue.mockImplementationOnce(() => Promise.resolve());

      await verifyStatusAndBody(
        await DELETE(buildDeleteNextRequest(MOCK_SUBSCRIPTION_3)),
        200,
        false,
      );
      expect(mockedDeleteSubscription.mock.calls.length).toStrictEqual(1);
      expect(mockedRemoveTargetAndDeleteRule.mock.calls.length).toStrictEqual(
        1,
      );
      expect(mockedDeleteQueue.mock.calls.length).toStrictEqual(1);
    });

    test("returns 200 on subscription with no targetId and queue", async () => {
      mockSettings(200);
      mockedDeleteSubscription.mockImplementationOnce(() => Promise.resolve());

      await verifyStatusAndBody(
        await DELETE(buildDeleteNextRequest(MOCK_SUBSCRIPTION_4)),
        200,
        false,
      );
      expect(mockedDeleteSubscription.mock.calls.length).toStrictEqual(1);
      expect(mockedRemoveTargetAndDeleteRule.mock.calls.length).toStrictEqual(
        0,
      );
      expect(mockedDeleteQueue.mock.calls.length).toStrictEqual(0);
    });

    test("returns 500 on AWS resource deletion failure", async () => {
      mockSettings(200);
      mockedRemoveTargetAndDeleteRule.mockRejectedValue({
        status: 500,
      });

      await verifyStatusAndBody(
        await DELETE(buildDeleteNextRequest(MOCK_SUBSCRIPTION_3)),
        500,
        false,
      );
      expect(mockedDeleteSubscription.mock.calls.length).toStrictEqual(0);
      expect(mockedRemoveTargetAndDeleteRule.mock.calls.length).toStrictEqual(
        1,
      );
      expect(mockedDeleteQueue.mock.calls.length).toStrictEqual(0);
    });

    test("returns 503 on successful aws resource deletion and unsuccessful subscription deletion", async () => {
      mockSettings(200);
      mockedDeleteSubscription.mockRejectedValue({
        status: 503,
      });
      mockedRemoveTargetAndDeleteRule.mockImplementationOnce(() =>
        Promise.resolve(),
      );
      mockedDeleteQueue.mockImplementationOnce(() => Promise.resolve());

      await verifyStatusAndBody(
        await DELETE(buildDeleteNextRequest(MOCK_SUBSCRIPTION_3)),
        503,
        false,
      );
      expect(mockedDeleteSubscription.mock.calls.length).toStrictEqual(1);
      expect(mockedRemoveTargetAndDeleteRule.mock.calls.length).toStrictEqual(
        1,
      );
      expect(mockedDeleteQueue.mock.calls.length).toStrictEqual(1);
    });

    test("returns 500 on settings failure", async () => {
      mockSettings(400);
      await verifyStatusAndBody(
        await DELETE(buildDeleteNextRequest(MOCK_SUBSCRIPTION_3)),
        500,
        false,
      );
    });

    test("returns 400 on invalid subscription in body", async () => {
      mockSettings(200);
      await verifyStatusAndBody(
        await DELETE(buildDeleteNextRequest({})),
        400,
        false,
      );
    });

    test("returns 400 on invalid json in body", async () => {
      mockSettings(200);
      await verifyStatusAndBody(
        await DELETE(buildDeleteNextRequest("")),
        400,
        false,
      );
    });
  });

  describe("Test for the POST subscriptions API", () => {
    beforeEach(() => {
      fetch.resetMocks();
      jest.clearAllMocks();
    });

    afterEach(() => {
      reset(mockReadOnlyHeaders);
    });

    test("returns 200 on successful case from existing destination", async () => {
      mockSettings(200);
      mockHeaders(LISC_NOTIFICATION);
      mockedGetEventBridgeDestination.mockResolvedValue([MOCK_DESTINATION]);
      mockedCreateEventBridgeResources.mockImplementationOnce(() =>
        Promise.resolve(),
      );
      mockedCreateSubscription.mockResolvedValue(
        MOCK_CREATE_SUBSCRIPTION_SP_API_RESPONSE,
      );

      await verifyStatusAndBody(await POST(), 200, true);
      expect(mockedGetEventBridgeDestination.mock.calls.length).toStrictEqual(
        1,
      );
      expect(
        mockedCreateEventBridgeDestination.mock.calls.length,
      ).toStrictEqual(0);
      expect(mockedCreateEventBridgeResources.mock.calls.length).toStrictEqual(
        1,
      );
      expect(mockedCreateSubscription.mock.calls.length).toStrictEqual(1);
    });

    test("returns 200 on successful case from create new destination", async () => {
      mockSettings(200);
      mockHeaders(LISC_NOTIFICATION);
      mockedGetEventBridgeDestination.mockResolvedValue([]);
      mockedCreateEventBridgeDestination.mockResolvedValue(
        MOCK_CREATE_DESTINATION_RESPONSE,
      );
      mockedCreateEventBridgeResources.mockImplementationOnce(() =>
        Promise.resolve(),
      );
      mockedCreateSubscription.mockResolvedValue(
        MOCK_CREATE_SUBSCRIPTION_SP_API_RESPONSE,
      );

      await verifyStatusAndBody(await POST(), 200, true);

      expect(mockedGetEventBridgeDestination.mock.calls.length).toStrictEqual(
        1,
      );
      expect(
        mockedCreateEventBridgeDestination.mock.calls.length,
      ).toStrictEqual(1);
      expect(mockedCreateEventBridgeResources.mock.calls.length).toStrictEqual(
        1,
      );
      expect(mockedCreateSubscription.mock.calls.length).toStrictEqual(1);
    });

    test("returns 500 on settings failure", async () => {
      mockSettings(400);

      await verifyStatusAndBody(await POST(), 500, false);
      expect(mockedGetEventBridgeDestination.mock.calls.length).toStrictEqual(
        0,
      );
      expect(
        mockedCreateEventBridgeDestination.mock.calls.length,
      ).toStrictEqual(0);
      expect(mockedCreateEventBridgeResources.mock.calls.length).toStrictEqual(
        0,
      );
      expect(mockedCreateSubscription.mock.calls.length).toStrictEqual(0);
    });

    test("returns 400 exception from missing Notification Type header from Create Subscription API", async () => {
      mockSettings(200);
      mockHeaders(undefined);

      await verifyStatusAndBody(await POST(), 400, false);
      expect(mockedGetEventBridgeDestination.mock.calls.length).toStrictEqual(
        0,
      );
      expect(
        mockedCreateEventBridgeDestination.mock.calls.length,
      ).toStrictEqual(0);
      expect(mockedCreateEventBridgeResources.mock.calls.length).toStrictEqual(
        0,
      );
      expect(mockedCreateSubscription.mock.calls.length).toStrictEqual(0);
    });

    test("returns 400 on 400 from GetEventBridge Destination call when invoking the CreateSubscription call", async () => {
      mockSettings(200);
      mockHeaders(LISC_NOTIFICATION);
      mockedGetEventBridgeDestination.mockRejectedValue({
        status: 400,
      });

      await verifyStatusAndBody(await POST(), 400, true);
      expect(mockedGetEventBridgeDestination.mock.calls.length).toStrictEqual(
        1,
      );
      expect(
        mockedCreateEventBridgeDestination.mock.calls.length,
      ).toStrictEqual(0);
      expect(mockedCreateEventBridgeResources.mock.calls.length).toStrictEqual(
        0,
      );
      expect(mockedCreateSubscription.mock.calls.length).toStrictEqual(0);
    });

    test("returns 400 on 400 from invoking CreateSubscription API", async () => {
      mockSettings(200);
      mockHeaders(LISC_NOTIFICATION);
      mockedGetEventBridgeDestination.mockResolvedValue([]);
      mockedCreateEventBridgeDestination.mockResolvedValue(
        MOCK_CREATE_DESTINATION_RESPONSE,
      );
      mockedCreateEventBridgeResources.mockImplementationOnce(() =>
        Promise.resolve(),
      );
      mockedCreateSubscription.mockRejectedValue({
        status: 400,
      });

      await verifyStatusAndBody(await POST(), 400, true);
      expect(mockedGetEventBridgeDestination.mock.calls.length).toStrictEqual(
        1,
      );
      expect(
        mockedCreateEventBridgeDestination.mock.calls.length,
      ).toStrictEqual(1);
      expect(mockedCreateEventBridgeResources.mock.calls.length).toStrictEqual(
        0,
      );
      expect(mockedCreateSubscription.mock.calls.length).toStrictEqual(1);
    });

    test("returns 200 for 409 response from CreateSubscription API", async () => {
      mockSettings(200);
      mockHeaders(LISC_NOTIFICATION);
      mockedGetEventBridgeDestination.mockResolvedValue([]);
      mockedCreateEventBridgeDestination.mockResolvedValue(
        MOCK_CREATE_DESTINATION_RESPONSE,
      );
      mockedCreateEventBridgeResources.mockImplementationOnce(() =>
        Promise.resolve(),
      );
      mockedCreateSubscription.mockRejectedValue({
        status: 409,
      });

      await verifyStatusAndBody(await POST(), 200, true);

      expect(mockedGetEventBridgeDestination.mock.calls.length).toStrictEqual(
        1,
      );
      expect(
        mockedCreateEventBridgeDestination.mock.calls.length,
      ).toStrictEqual(1);
      expect(mockedCreateEventBridgeResources.mock.calls.length).toStrictEqual(
        0,
      );
      expect(mockedCreateSubscription.mock.calls.length).toStrictEqual(1);
    });
  });
});
