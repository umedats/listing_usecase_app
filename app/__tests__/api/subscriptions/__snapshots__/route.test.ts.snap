// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the subscriptions API Test for the GET subscriptions API returns 200 on successful case 1`] = `
{
  "data": [
    {
      "eventBusFilterRule": "SellingPartnerNotificationsAPIRule",
      "eventBusName": "TestEventBus",
      "notificationType": "BRANDED_ITEM_CONTENT_CHANGE",
      "sqsQueueName": "BrandedItemContentChange",
      "subscriptionId": "f8c3de3d-1fea-4d7c-a8b0-29f63c4c3454",
    },
    {
      "eventBusFilterRule": "SellingPartnerNotificationsAPIRule",
      "eventBusName": "TestEventBus1",
      "notificationType": "PRODUCT_TYPE_DEFINITIONS_CHANGE",
      "sqsQueueName": "ProductTypeDefinitionsChange",
      "subscriptionId": "510f082f-a853-4847-acf2-475641cb7b06",
    },
  ],
  "debugContext": [],
}
`;

exports[`Test for the subscriptions API Test for the GET subscriptions API returns 409 on runtime error 1`] = `
{
  "debugContext": [],
}
`;

exports[`Test for the subscriptions API Test for the POST subscriptions API returns 200 for 409 response from CreateSubscription API 1`] = `
{
  "data": {},
  "debugContext": [],
}
`;

exports[`Test for the subscriptions API Test for the POST subscriptions API returns 200 on successful case from create new destination 1`] = `
{
  "debugContext": [],
}
`;

exports[`Test for the subscriptions API Test for the POST subscriptions API returns 200 on successful case from existing destination 1`] = `
{
  "debugContext": [],
}
`;

exports[`Test for the subscriptions API Test for the POST subscriptions API returns 400 on 400 from GetEventBridge Destination call when invoking the CreateSubscription call 1`] = `
{
  "debugContext": [],
}
`;

exports[`Test for the subscriptions API Test for the POST subscriptions API returns 400 on 400 from invoking CreateSubscription API 1`] = `
{
  "debugContext": [],
}
`;
